File: customer-address-jpa/src/main/java/com/izertis/example/service/impl/CustomerServiceImpl.java
package com.izertis.example.service.impl;

import com.izertis.example.domain.Customer;
import com.izertis.example.events.CustomerEventsProducer;
import com.izertis.example.service.impl.mappers.CustomerServiceMapper;
import com.izertis.example.service.impl.mappers.EventsMapper;
import com.izertis.example.service.CustomerService;
import com.izertis.example.service.dtos.CustomerSearchCriteria;
import com.izertis.example.repository.jpa.CustomerRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

/** Service Implementation for managing [Customer]. */
@Service
@Transactional(readOnly = true)
@lombok.AllArgsConstructor
public class CustomerServiceImpl implements CustomerService {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private final CustomerServiceMapper customerServiceMapper = CustomerServiceMapper.INSTANCE;

    private final CustomerRepository customerRepository;

    private final EventsMapper eventsMapper = EventsMapper.INSTANCE;

    private final CustomerEventsProducer eventsProducer;

    private final ApplicationEventPublisher applicationEventPublisher;

    @Transactional
    public Customer createCustomer(Customer input) {
        log.debug("[CRUD] Request to save Customer: {}", input);
        var customer = customerServiceMapper.update(new Customer(), input);
        customer = customerRepository.save(customer);
        // TODO: may need to reload the entity to fetch relationships 'mapped by id'
        // emit events
        var customerEvent = eventsMapper.asCustomerEvent(customer);
        eventsProducer.onCustomerEvent(customerEvent);
        return customer;
    }

    public Optional<Customer> getCustomer(Long id) {
        log.debug("[CRUD] Request to get Customer : {}", id);
        var customer = customerRepository.findById(id);
        return customer;
    }

    @Transactional
    public Optional<Customer> updateCustomer(Long id, Customer input) {
        log.debug("Request updateCustomer: {} {}", id, input);

        var customer = customerRepository.findById(id).map(existingCustomer -> {
            return customerServiceMapper.update(existingCustomer, input);
        }).map(customerRepository::save);
        if (customer.isPresent()) {
            // emit events
            var customerEvent = eventsMapper.asCustomerEvent(customer.get());
            eventsProducer.onCustomerEvent(customerEvent);
        }
        return customer;
    }

    @Transactional
    public void deleteCustomer(Long id) {
        log.debug("[CRUD] Request to delete Customer : {}", id);
        customerRepository.deleteById(id);
        // emit events
        var customerEvent = eventsMapper.asCustomerEvent(id);
        eventsProducer.onCustomerDeletedEvent(customerEvent);
    }

    public Page<Customer> searchCustomers(CustomerSearchCriteria input, Pageable pageable) {
        log.debug("Request searchCustomers: {} {}", input, pageable);

        var customers = customerRepository.findAll(pageable);
        return customers;
    }

}

File: kustomer-address-jpa/src/main/kotlin/com/izertis/example/service/impl/CustomerServiceImpl.kt
package com.izertis.example.service.impl

// import com.izertis.examples.events.dtos.*

import com.izertis.example.domain.Customer
import com.izertis.example.events.CustomerEventsProducer
import com.izertis.example.repository.jpa.CustomerRepository
import com.izertis.example.service.CustomerService
import com.izertis.example.service.dtos.CustomerSearchCriteria
import com.izertis.example.service.impl.mappers.CustomerServiceMapper
import com.izertis.example.service.impl.mappers.EventsMapper
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/** Service Implementation for managing [com.izertis.example.domain.Customer]. */
@Service
@Transactional(readOnly = true)
open class CustomerServiceImpl(
    private val customerRepository: CustomerRepository,
    private val eventsProducer: CustomerEventsProducer
) : CustomerService {

  private val log: Logger = LoggerFactory.getLogger(javaClass)

  private val customerServiceMapper: CustomerServiceMapper = CustomerServiceMapper.INSTANCE

  private val eventsMapper: EventsMapper = EventsMapper.INSTANCE

  @Transactional
  override fun createCustomer(input: Customer): Customer {
    log.debug("[CRUD] Request to save Customer: {}", input)
    return customerServiceMapper.update(Customer(), input).let(customerRepository::save).also {
      eventsProducer.onCustomerEvent(eventsMapper.asCustomerEvent(it))
    }
  }

  override fun getCustomer(id: Long): Customer? {
    log.debug("[CRUD] Request to get Customer : {}", id)
    val customer = customerRepository.findByIdOrNull(id)
    return customer
  }

  @Transactional
  override fun updateCustomer(id: Long, input: Customer): Customer? {
    log.debug("Request updateCustomer: {} {}", id, input)

    return customerRepository
        .findByIdOrNull(id)
        ?.let { customerServiceMapper.update(it, input) }
        ?.let { customerRepository.save(it) }
        ?.also { eventsProducer.onCustomerEvent(eventsMapper.asCustomerEvent(it)) }
  }

  @Transactional
  override fun deleteCustomer(id: Long): Unit {
    log.debug("[CRUD] Request to delete Customer : {}", id)
    customerRepository.deleteById(id)
    eventsProducer.onCustomerEvent(eventsMapper.asCustomerEvent(id))
  }

  override fun searchCustomers(input: CustomerSearchCriteria, pageable: Pageable): Page<Customer> {
    log.debug("Request searchCustomers: {} {}", input, pageable)
    val customers = customerRepository.findAll(pageable)
    return customers
  }
}
